### PR TITLE
Move metadata class model de/serialization to sub-package

### PR DESCRIPTION
**Pull request by lukpueh**
Fixes #-
Related to #1270 

**Description of the changes being introduced by the pull request**:
This PR moves de/serialization logic of TUF metadata from the metadata class model (`api.metadata`) to a newly added serialization sub-package (`api.serialization`) and adopts the metadata API accordingly, with the goal of making it easier for users to swap de/serialization implementations. In particular this PR adds:
- abstract base classes to serialize and deserialize to and from wireline metadata bytes (for transport), plus default json implementations, and
- an abstract base class to serialize signed bytes (for signature creation and verification) and a default canonical json implementation.

*More infos can be found in ...*
- #1270, which drafts a corresponding ADR and discusses the pros and cons of this and alternative approaches,
-  79ce0346, whose commit message has very specific design considerations, and
- the other commit messages, which are informative too. ;)

**Discussion:**

The meat of the default json and canonical json de/serializers is the `from/to_dict` logic, which is currently implemented in a `serialization.util` module. I wonder if we shouldn't just keep it on the metadata class model instead (we'd only need to revert b786bc0).

*Pro `from/to_dict` as metadata class methods:*
   - better structured because encapsulated on each class
   - keeps the serialization package minimal
   - `from/to_dict` is hardly serialization, which implies bytes, but rather some sort of type casting / conversion, which seems appropriate to be implemented on the class
   - it would spare us the stupid cyclic imports
   
*Pro `*_from/to_dict` in serialization sub-package:*
   - might make it easier to copy-paste-customize when creating a custom de/serializer implementation.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]:
```suggestion
            serializer = JSONSerializer(compact=True)
```
Using the named argument makes it clearer what's going on and obviates the need for the comment.

File: tuf/api/serialization/util.py
Comment (MVrachev) [inline tuf/api/serialization/util.py]: How did you decide to accept `typing.Mapping` instead of `typing.Dict` here and elsewhere in this file?

Comment (lukpueh) [reply tuf/api/serialization/util.py]: Good question! The [`typing.Dict` docs](https://docs.python.org/3/library/typing.html#typing.Dict) have the answer:
> To annotate arguments it is preferred to use an abstract collection type such as Mapping.

I didn't really question that advice, but it makes sense when you consider the API design principle of returning specific types, while accepting generic types.

File: tuf/api/pylintrc
Comment (joshuagl) [inline tuf/api/pylintrc:12]: I'm not sure what change in this first commit (79ce03467cf480f5922b2fa3de85a83b766a5e29) necessitated this?  Do we want to make this decision for the entirety of tuf/api? Or should we disable this on a case-by-case basis (at the class or module level)?

File: tuf/api/serialization/__init__.py
Comment (joshuagl) [inline tuf/api/serialization/__init__.py:4]: Should we add SPDX style license headers to new files?
```
# Copyright New York University and the TUF contributors
# SPDX-License-Identifier: MIT OR Apache-2.0
```

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py:4]: Same comment about SPDX license headers:
```
# Copyright New York University and the TUF contributors
# SPDX-License-Identifier: MIT OR Apache-2.0
```

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Note: the recently adopted style guide recommends _against_ importing individual classes or functions.

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Nit: could we rename this variable, rather than using the `_` trick? Maybe json_dict, json_object (as the return type of `json.loads` is a Python object), or loaded_json.

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Out of curiosity, why did you opt for 1 space indent for the non-compact version?

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: What happened here? :-)

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Should the typing annotation mark deserializer as optional, like storage_backend ?

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: As above, I think the type annotations for serializer and storage_backend should mark them as optional

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Also here, serializer is optional.

Nit: I wonder if we should rename the parameter signed_serializer. Using the same name for named parameters that take different types could be confusing.

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Same comments as above, serializer is optional and might be better named signed_serializer

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Same comment here, with the code moved, can we rename _dict? Possible alternatives include  json_object and json_dict.

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Are we aiming for three blank lines between top-level definitions? The style guide suggests two: https://google.github.io/styleguide/pyguide.html#35-blank-lines

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Same comment on blank lines between top-level definitions.

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py]: Same blank lines comment here, too.

**Review by joshuagl — COMMENTED** [review]
Lovely PR, apologies for the delay in getting to review this @lukpueh.

I have a few minor questions and suggestions in-line; mostly around typing annotations, and variable & parameter names. 

I also want to point out that the new code we are writing since introducing a new coding style seems to consistently violate the style guide's recommendations around imports, [specifically](https://google.github.io/styleguide/pyguide.html#22-imports):

> Use `import` statements for packages and modules only, not for individual classes or functions.

New code in tuf.api is fairly consistently importing individual classes and functions from modules. Should we consider whether we want to make a style guide modification at this point?

> * [79ce034](https://github.com/theupdateframework/tuf/commit/79ce03467cf480f5922b2fa3de85a83b766a5e29), whose commit message has very specific design considerations

Thank you for the detailed commit messages. I wonder whether we might want to capture some of this detail in the ADR?

> **Discussion:**
[snip]
> _Pro `from/to_dict` as metadata class methods:_
> 
> * better structured because encapsulated on each class
> * keeps the serialization package minimal
> * `from/to_dict` is hardly serialization, which implies bytes, but rather some sort of type casting / conversion, which seems appropriate to be implemented on the class
> * it would spare us the stupid cyclic imports

Having reviewed this PR, I am in favour of keeping `[from|to]_dict` as metadata class methods. Curious to read what others think.

File: tuf/api/pylintrc
Comment (lukpueh) [reply tuf/api/pylintrc:12]: > I'm not sure what change in this first commit (79ce034) necessitated this?

The admittedly long commit message has an admittedly subtle hint:
```
 - configure api/pylintrc to allow classes without public methods (default was 2)
```

In other words, all the abstract classes in `serialization/__init__.py ` are one method short of the pylint ideal.

> Do we want to make this decision for the entirety of tuf/api? Or should we disable this on a case-by-case basis (at the class or module level)?

The check seems a bit controversial (see stackoverflow and pylint issue tracker). So I think we can just disable it.

File: tuf/api/serialization/json.py
Comment (lukpueh) [reply tuf/api/serialization/json.py]: I know. I guess I just kept using the convention that we already used in this file, which we started before fully embracing our new style guide... and because PEP 8 says it's okay.

File: tuf/api/serialization/json.py
Comment (lukpueh) [reply tuf/api/serialization/json.py]: I just adopted what was already there:
https://github.com/theupdateframework/tuf/blob/74b1549c0041fd6e13c0277713803dd61d2baca4/tuf/repository_lib.py#L444-L445

File: tuf/api/serialization/util.py
Comment (trishankatdatadog) [inline tuf/api/serialization/util.py]: Whoa. Clever, but maybe a bit too much magic?

Comment (lukpueh) [reply tuf/api/serialization/util.py]: Yes probably. :'(

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: I must admit that I follow this secret personal convention:

Based on the guide I use "single blank lines as [I] judge appropriate within functions or methods",  but then I use two blank lines between methods, to better distinguish from the single ones, and then three blank lines between top-level definitions, to distinguish from the two blank lines.

I should probably either strictly adhere to the guide, or propose a style guide amendment.

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: See https://github.com/theupdateframework/tuf/pull/1279#discussion_r582149810

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: See https://github.com/theupdateframework/tuf/pull/1279#discussion_r582149810

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: See https://github.com/theupdateframework/tuf/pull/1279#discussion_r582149810

**Comment by lukpueh** [comment]
Thanks for the detailed review and great suggestions, @joshuagl! I left a couple of replies inline and will address the rest in code. Regarding...

> New code in tuf.api is fairly consistently importing individual classes and functions from modules. Should we consider whether we want to make a style guide modification at this point?

Maybe we should ... also in the light of #1261. Do you think we should get this straight before landing this PR?

> Thank you for the detailed commit messages. I wonder whether we might want to capture some of this detail in the ADR?

I tried to capture the gist without making the ADR too verbose. But I'll try to revise the ADR once more. :) 

> Having reviewed this PR, I am in favour of keeping `[from|to]_dict` as metadata class methods. Curious to read what others think.

Cool. I'll wait just a bit longer to see if others still want to weigh in.

**Comment by trishankatdatadog** [comment]
> Having reviewed this PR, I am in favour of keeping `[from|to]_dict` as metadata class methods. Curious to read what others think.

Can someone enlighten me as to why? Cos I don't see it yet.

**Comment by MVrachev** [comment]
> > Having reviewed this PR, I am in favour of keeping `[from|to]_dict` as metadata class methods. Curious to read what others think.
> 
> Can someone enlighten me as to why? Cos I don't see it yet.

My personal feeling is that `from_dict`  functions are logical to be located inside the classes because they create a new instance.
I see them as a second constructor. 
About `to_dict` methods I see them as `casting` which again feels like it's a functionality logically connected with the classes.

**Comment by trishankatdatadog** [comment]
@MVrachev but are they useful to other (de)serializers?

**Comment by MVrachev** [comment]
> @MVrachev but are they useful to other (de)serializers?

I am not sure where are the `[from|to]_dict` functions located relate to the other (de)serializers.
Could you elaborate more on this @trishankatdatadog?

**Comment by jku** [comment]
> > @MVrachev but are they useful to other (de)serializers?
> 
> I am not sure where are the `[from|to]_dict` functions located relate to the other (de)serializers.
> Could you elaborate more on this @trishankatdatadog?

The point Trishank is making is that they look like implementation details of the json de/serializer hidden inside metadata.py. And I think we should admit that that's what they are -- e.g. `from_dict()` is not a generic constructor it's clearly a helper for deserializing json.

On the other hand, the reason Lukas is proposing we leave them in is that it makes the code simpler -- I agree with this as well: the compromise seems reasonable and should not make other de/serializers any harder to implement.

**Comment by lukpueh** [comment]
FYI, just left another comment on https://github.com/theupdateframework/tuf/pull/1270#issuecomment-785965015

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py]: Agreed, we should strictly adhere to the guide or propose an amendment. I'm neither for or against this personal convention, but I do think amendments to the style guide which contradict the "upstream" Google guide should be kept to a minimum.

File: tuf/api/serialization/util.py
Comment (joshuagl) [reply tuf/api/serialization/util.py]: Agreed, the if/else cascade is uglier but less magical and easier to follow.

**Comment by joshuagl** [comment]
> Thanks for the detailed review and great suggestions, @joshuagl! I left a couple of replies inline and will address the rest in code. Regarding...
> 
> > New code in tuf.api is fairly consistently importing individual classes and functions from modules. Should we consider whether we want to make a style guide modification at this point?
> 
> Maybe we should ... also in the light of #1261. Do you think we should get this straight before landing this PR?

No need to block this PR. However, I think we should figure this out and ensure the new code conforms to the style ASAP. Otherwise we risk similar incidents unintentionally contradicting the style guide in future. At the same time as we ensure new code conforms to the style, it would be good to add style checking to CI – and perhaps have a [black](https://black.readthedocs.io/en/stable/) configuration for folks to ensure their code matches before submission.

> The point Trishank is making is that they look like implementation details of the json de/serializer hidden inside metadata.py. And I think we should admit that that's what they are -- e.g. from_dict() is not a generic constructor it's clearly a helper for deserializing json.

> On the other hand, the reason Lukas is proposing we leave them in is that it makes the code simpler -- I agree with this as well: the compromise seems reasonable and should not make other de/serializers any harder to implement.

This matches my understanding, and opinion, also. The compromise is pragmatic and reasonable, but let's be sure the comments indicate that those methods are a helper primarily for conversion to/from JSON.

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py]: You are right. Let's play by the rules we chose. :) Sorry for the noise.

File: tox.ini
Comment (MVrachev) [inline tox.ini:48]: Is there a way to ignore all files under a certain folder?
That way you would be able to ignore everything under `tuf/api`.

File: tuf/api/serialization/json.py
Comment (MVrachev) [reply tuf/api/serialization/json.py]: I agree with Joshua on this one. 
We don't have a linter that would automatically catch those mistakes, but it's important we correct each other.
Also, it seems logical to forbid importing of functions and classes.
This way we prevent future problems with naming and namespaces.

File: tox.ini
Comment (lukpueh) [reply tox.ini:48]: > Is there a way to ignore all files under a certain folder?

Nope (see my comment). I even looked at the implementation of  `--ignore` and `--ignore-patterns`.

Comment (lukpueh) [reply tox.ini:48]: IIRC there's a corresponding ticket on the pylint issue tracker.

**Comment by lukpueh** [comment]
This could use another review. :)

- Following discussions here, in #1270, and elsewhere I reverted the commit that moved to/from_dict methods to a separate module (8e9afc96).
- Unfortunately, I was wrong that this frees us from cyclic imports, but I at least updated the corresponding code comments (88582800).
- Given that to/from_dict remain on the class, I made an effort to demystify the design  of the subclassed 'Signed' class by adding more explicit helpers (c254fd04).
- And I tried to largely remove JSON-bias from the metadata class model so that users understand that JSON is only (the default) one of many TUF wireline formats (f421cdce).

- Apart from that I addressed code review comments regarding license boilerplate (2f57eb8e), the Optional type hint  (71a684b5), a named argument (0b92da91), argument renames (b934dedd) and blank lines (17c25e29), and a linter warning about inconsistent returns (8be8ec1c).

File: tuf/api/pylintrc
Comment (joshuagl) [reply tuf/api/pylintrc:12]: Thanks for explaining.

File: tuf/api/serialization/__init__.py
Comment (joshuagl) [inline tuf/api/serialization/__init__.py]: Note: now that we are only supporting Python3 we can switch to less confusing metaclass syntax, either
```Python
import abc

class MetadataDeserializer(ABC):
```
or
```Python
import abc

class MetadataDeserializer(metaclass=ABCMeta):
```
we can handle this in a future PR, though.

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py:70]: Do you think it is worth clarifying, via the class name, that this is the OLPC canonical JSON?
You mention it in the header above, but as we have had several adopters surprised that our Canonical JSON isn't compatible with go-tuf's Canonical JSON, I wonder whether it is worth being very explicit?

Counter-argument, I do not like `OLPCCanonicalJSONSerializer` as a class name...

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Why are we using six in code that was never intended to support Python2?

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Python3 syntax for this appears to be 
```Python
            raise SerializationError from e
```

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Let's not use six here
```Python
            raise SerializationError from e
```

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py:206]: I think we're missing some of the `serializer` -> `signed_serializer` rename in this patch (885828007dd5b5a68826470dce3dbe123d065aab)? The argument name and the variable assigned below are still `serializer`

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py:320]: Should this have the `@staticmethod` decorator also? Throughout the rest of this patch it is not being called on an instance of the class, but on the `super()` proxy object, much like `_common_fields_from_dict()`.

File: tuf/api/serialization/json.py
Comment (joshuagl) [inline tuf/api/serialization/json.py]: Python3 syntax for this appears to be 
```Python
            raise SerializationError from e
```

**Review by joshuagl — CHANGES_REQUESTED** [review]
Thank you for the updates Lukas. I have a couple of minor nits and one surprise, why are we using six in code that was never intended to run on Python2?

File: tuf/api/serialization/json.py
Comment (lukpueh) [reply tuf/api/serialization/json.py:70]: I agree with both arguments. :D

File: tuf/api/serialization/json.py
Comment (lukpueh) [reply tuf/api/serialization/json.py]: For `raise_from` below, which I will change to `raise ... from`.

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py:206]: It's the other way around, this line should have gone into b934deddd. Apologies for the bad `git add --patch`ing. Happy to rebase. Maybe after a final LGTM?

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py:320]: Can't be static because it needs `self`, but thanks for pointing out that those two are called on `super()`. I guess it's less confusing to call them on `self`, given that they are only called in the parent class. 🤷

File: tuf/api/serialization/json.py
Comment (lukpueh) [reply tuf/api/serialization/json.py:70]: Another way to avoid ambiguity would be to not call it `CanonicalJSONSerializer` at all, but something like `DefaultSignedSerializer` or so? Then people are forced to read the docstring to get more info. At any rate, I will mention OLPC in the class and function docstrings.

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py:206]: WFM

File: tuf/api/metadata.py
Comment (joshuagl) [reply tuf/api/metadata.py:320]: Sounds good, thanks.

**Comment by lukpueh** [comment]
Thanks for the review! I (hope I) addressed all your comments in the last 3 commits. I chose to leave `CanonicalJSONSerializer` as is but update the docstrings. Let me know if that works for you.

I'm happy to also fix the minor git history hiccup from https://github.com/theupdateframework/tuf/pull/1279#discussion_r590493044 once you have approved the PR.

**Review by joshuagl — APPROVED** [review]
Thank you for quickly addressing the review comments.

> Thanks for the review! I (hope I) addressed all your comments in the last 3 commits. I chose to leave `CanonicalJSONSerializer` as is but update the docstrings. Let me know if that works for you.

Last 3 commits look great, thanks. I'm happy to leave `CanonicalJSONSerializer`. Let's see if we can come up with a 'better' name before we release with an API stability promise.

> I'm happy to also fix the minor git history hiccup from [#1279 (comment)](https://github.com/theupdateframework/tuf/pull/1279#discussion_r590493044) once you have approved the PR.

That would be good, thanks!

**Comment by lukpueh** [comment]
> > I'm happy to also fix the minor git history hiccup from [#1279 (comment)](https://github.com/theupdateframework/tuf/pull/1279#discussion_r590493044) once you have approved the PR.
> 
> That would be good, thanks!

Done. Given that the [overall diff of the PR didn't change](https://github.com/theupdateframework/tuf/compare/a223961..ef91964) I suppose your approval is still valid. Merging...

**Pull request merged**
